tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -14211,7 +14357,82 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
         // HandleDeclAttributeForHLSL()
         pMaxDispatchGrid = pAttr;
       }
+      break; 
+    case AttributeList::AT_HLSLNumThreads: {
+      const auto *SM =
+          hlsl::ShaderModel::GetByName(getLangOpts().HLSLProfile.c_str());
+
+      const std::string &entryName = getLangOpts().HLSLEntryFunction;
+
+      bool isEntry = false;
+      bool isNode = false;
+      bool isCS = false;
+      bool isMS = false;
+      bool isAS = false;
+
+      isEntry = !getLangOpts().IsHLSLLibrary &&
+                DC->getDeclKind() == Decl::Kind::TranslationUnit &&
+                functionName == entryName;
+      AttributeList *pAL = D.getDeclSpec().getAttributes().getList();
+      while (pAL) {
+        if (pAL->getKind() == AttributeList::AT_HLSLShader) {
+          Expr *ArgExpr = pAL->getArgAsExpr(0);
+          StringLiteral *Literal =
+              dyn_cast<StringLiteral>(ArgExpr->IgnoreParenCasts());
+          DXIL::ShaderKind Stage =
+              ShaderModel::KindFromFullName(Literal->getString());
+          isNode |= Stage == DXIL::ShaderKind::Node;
+          isCS |= Stage == DXIL::ShaderKind::Compute;
+          isMS |= Stage == DXIL::ShaderKind::Mesh;
+          isAS |= Stage == DXIL::ShaderKind::Amplification;
+        }
+        pAL = pAL->getNext();
+      }
+
+      if (!isNode && SM->IsLib() &&
+          ((!isCS && !isMS && !isAS) &&
+           (!SM->IsCS() && !SM->IsMS() && !SM->IsAS()))) {
```

Comment:
This looks similar to the other case, except for here we're checking both the local variable and the values from `SM` - should they be doing the same thing as each other? Furthermore this is a pretty difficult condition to read, it'd be nice to simplify it.

Can we consolidate the logic to work out which shader stage it is into a function and simplify all of this?

Reply:
They should not be doing the same thing. For example, isCS means that the function has a [shader("compute")] shader stage attribute. SM->isCS() means that the command line has information that the compilation is targeted to a compute profile, like -T cs_6_6. Glad you pointed this out though, because SM->isLib() actually implies the last 3 conditions, so I can just remove them!